### PR TITLE
 add permission check for bulkCreate when overwrite

### PR DESCRIPTION
**Pull request by raintygao**
### Description

 add permission check for bulkCreate when overwrite

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 1]
```diff
@@ -250,16 +251,40 @@ export class WorkspaceSavedObjectsClientWrapper {
...
+        throw generateSavedObjectsPermissionError();
+      }
+
+      if (options.overwrite) {
+        for (const object of objects) {
+          const { type, id } = object;
+          if (
+            id &&
+            !(await this.validateWorkspacesAndSavedObjectsPermissions(
+              await wrapperOptions.client.get(type, id),
```
Comment (wanglam) [inline src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:273]: @SuZhou-Joe What do you think  about if we need to check passed `object.id` was exists?

Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:273]: @SuZhou-Joe  Could you take a look? Do you have any comments?

Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:273]: If no id is provided, the bulkCreate will create brand new object inside `options.workspaces`. So I'd recommend that.
1. If id exists, do the permission check shown above.
2. If id not exist, check the library_write permission on the `options.workspaces`.

Comment (wanglam) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:273]: I think we need to check if the object of provided id exists first. If we can't get this object, I think we can ignore the permission validation here. This case is for bulkCreate with override option, but passed some not exists objects. We need to skip permission validation for these objects. Or the bulkCreate will throw an not found error.

Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:273]: > This case is for bulkCreate with override option, but passed some not exists objects. We need to skip permission validation for these objects. Or the bulkCreate will throw an not found error.

Would it be more valid if we put this logic in caller side or still in client wrapper ?

Comment (wanglam) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:273]: In my opinion, pass brand new objects here with override option is a valid case here. Throw object not found error will deny this behavior.

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 2]
```diff
@@ -250,16 +251,52 @@ export class WorkspaceSavedObjectsClientWrapper {
...
+          if (id) {
+            let rawObject;
+            try {
+              rawObject = await wrapperOptions.client.get(type, id);
+            } catch (error) {
+              // If object is not found, we will skip the validation of this object.
+              if (SavedObjectsErrorHelpers.isNotFoundError(error as Error)) {
+                continue;
+              } else {
+                throw generateWorkspacePermissionError();
```
Comment (wanglam) [inline src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:279]: I think we can just throw the `client.get` error like other places.

**Review by wanglam — APPROVED** [review]

**Pull request merged**